sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala

```diff
@@ -0,0 +1,367 @@
+  }
+
+  def executionMetrics(executionId: Long): Map[Long, String] = {
+    Option(liveExecutions.get(executionId)).map { exec =>
+      if (exec.metricsValues != null) {
+        exec.metricsValues
+      } else {
+        aggregateMetrics(exec)
+      }
+    }.getOrElse {
+      throw new NoSuchElementException(s"execution $executionId not found")
+    }
+  }
+
+  private def aggregateMetrics(exec: LiveExecutionData): Map[Long, String] = {
+    val metricIds = exec.metrics.map(_.accumulatorId).sorted
+    val metricTypes = exec.metrics.map { m => (m.accumulatorId, m.metricType) }.toMap
+    val metrics = exec.stages.toSeq
+      .flatMap { stageId => Option(stageMetrics.get(stageId)) }
+      .flatMap(_.taskMetrics.values().asScala)
+      .flatMap { metrics => metrics.ids.zip(metrics.values) }
+
+    val aggregatedMetrics = (metrics ++ exec.driverAccumUpdates.toSeq)
+      .filter { case (id, _) => metricIds.contains(id) }
+      .groupBy(_._1)
+      .map { case (id, values) =>
+        id -> SQLMetrics.stringValue(metricTypes(id), values.map(_._2).toSeq)
+      }
+
+    // Check the execution again for whether the aggregated metrics data has been calculated.
+    // This can happen if the UI is requesting this data, and the onExecutionEnd handler is
+    // running at the same time. The metrics calculcated for the UI can be innacurate in that
+    // case, since the onExecutionEnd handler will clean up tracked stage metrics.
+    if (exec.metricsValues != null) {
+      exec.metricsValues
+    } else {
+      aggregatedMetrics
+    }
+  }
+
+  private def updateStageMetrics(
+      stageId: Int,
+      attemptId: Int,
+      taskId: Long,
+      accumUpdates: Seq[AccumulableInfo],
+      succeeded: Boolean): Unit = {
+    Option(stageMetrics.get(stageId)).foreach { metrics =>
+      if (metrics.attemptId != attemptId || metrics.accumulatorIds.isEmpty) {
+        return
+      }
+
+      val oldTaskMetrics = metrics.taskMetrics.get(taskId)
+      if (oldTaskMetrics != null && oldTaskMetrics.succeeded) {
+        return
+      }
+
+      val updates = accumUpdates
+        .filter { acc => acc.update.isDefined && metrics.accumulatorIds.contains(acc.id) }
+        .sortBy(_.id)
+
+      if (updates.isEmpty) {
+        return
+      }
+
+      val ids = new Array[Long](updates.size)
+      val values = new Array[Long](updates.size)
+      updates.zipWithIndex.foreach { case (acc, idx) =>
+        ids(idx) = acc.id
+        // In a live application, accumulators have Long values, but when reading from event
+        // logs, they have String values. For now, assume all accumulators are Long and covert
+        // accordingly.
+        values(idx) = acc.update.get match {
+          case s: String => s.toLong
+          case l: Long => l
+          case o => throw new IllegalArgumentException(s"Unexpected: $o")
+        }
+      }
+
+      // TODO: storing metrics by task ID can lead to innacurate metrics when speculation is on.
```

Comment:
I think its more general than this.  I'd say

Since we store metrics by taskID, in a way we'll double-count the stage metrics when there are multiple tasks for a given index -- in particular, if there is speculation, or if there are multiple attempts for a task.